python/demo/demo_pml/demo_pml.py

```diff
@@ -0,0 +1,606 @@
+# $\varepsilon$ that takes the value $\varepsilon_m$
+# for cells inside the wire, while it takes the value of the
+# background permittivity $\varepsilon_b$ in the background region:
+
+D = fem.FunctionSpace(domain, ("DG", 0))
+eps = fem.Function(D)
+au_cells = cell_tags.find(au_tag)
+bkg_cells = cell_tags.find(bkg_tag)
+eps.x.array[au_cells] = np.full_like(
+    au_cells, eps_au, dtype=np.complex128)
+eps.x.array[bkg_cells] = np.full_like(bkg_cells, eps_bkg, dtype=np.complex128)
+eps.x.scatter_forward()
+
+# Now we need to define our weak form in DOLFINx.
+# Let's write the PML weak form first. As a first step,
+# we can define our new complex coordinates as:
+
+# +
+x = ufl.SpatialCoordinate(domain)
+alpha = 1
+
+# PML corners
+xy_pml = ufl.as_vector((pml_coordinates(x[0], alpha, k0, l_dom, l_pml),
+                        pml_coordinates(x[1], alpha, k0, l_dom, l_pml)))
+
+# PML rectangles along x
+x_pml = ufl.as_vector((pml_coordinates(x[0], alpha, k0, l_dom, l_pml), x[1]))
+
+# PML rectangles along y
+y_pml = ufl.as_vector((x[0], pml_coordinates(x[1], alpha, k0, l_dom, l_pml)))
+# -
+
+# We can then express this coordinate systems as
+# a material transformation within the PML region. In other words,
+# the PML region can be interpreted as a material
+# having, in general, anisotropic, inhomogeneous and complex permittivity
+# $\boldsymbol{\varepsilon}_{pml}$ and
+# permeability $\boldsymbol{\mu}_{pml}$. To do this, we need
+# to calculate the Jacobian of the coordinate transformation:
+#
+# $$
+# \mathbf{J}=\mathbf{A}^{-1}= \nabla\boldsymbol{x}^
+# \prime(\boldsymbol{x}) =
+# \left[\begin{array}{ccc}
+# \frac{\partial x^{\prime}}{\partial x} &
+# \frac{\partial y^{\prime}}{\partial x} &
+# \frac{\partial z^{\prime}}{\partial x} \\
+# \frac{\partial x^{\prime}}{\partial y} &
+# \frac{\partial y^{\prime}}{\partial y} &
+# \frac{\partial z^{\prime}}{\partial y} \\
+# \frac{\partial x^{\prime}}{\partial z} &
+# \frac{\partial y^{\prime}}{\partial z} &
+# \frac{\partial z^{\prime}}{\partial z}
+# \end{array}\right]=\left[\begin{array}{ccc}
+# \frac{\partial x^{\prime}}{\partial x} & 0 & 0 \\
+# 0 & \frac{\partial y^{\prime}}{\partial y} & 0 \\
+# 0 & 0 & \frac{\partial z^{\prime}}{\partial z}
+# \end{array}\right]=\left[\begin{array}{ccc}
+# J_{11} & 0 & 0 \\
+# 0 & J_{22} & 0 \\
+# 0 & 0 & 1
+# \end{array}\right]
+# $$
+#
+# Then, our $\boldsymbol{\varepsilon}_{pml}$ and
+# $\boldsymbol{\mu}_{pml}$ can be calculated with
+# the following formula, from
+# [Ward & Pendry, 1996](
+# https://www.tandfonline.com/doi/abs/10.1080/09500349608232782):
+#
+# $$
+# \begin{align}
+# & {\boldsymbol{\varepsilon}_{pml}} =
+# A^{-1} \mathbf{A} {\boldsymbol{\varepsilon}_b}\mathbf{A}^{T},\\
+# & {\boldsymbol{\mu}_{pml}} =
+# A^{-1} \mathbf{A} {\boldsymbol{\mu}_b}\mathbf{A}^{T},
+# \end{align}
+# $$
+#
+# with $A^{-1}=\operatorname{det}(\mathbf{J})$.
+#
+# In DOLFINx, we use
+# `ufl.grad` to calculate the Jacobian of our
```

Comment:
For UPML implementations `\varepsilon` and `\mu` are given by

![image](https://user-images.githubusercontent.com/206581/210114904-3909e432-f056-4cf2-b2b9-12f954e1d301.png)

(http://www.mit.edu/~wsshin/pdf/shin2012jcp.pdf)

I think you don't have to use `ufl.grad`, but it is nice to see that this is possible using the `ufl`.

Reply:
Did not know about this definition of uniaxial PML, thank you!

I am not familiar with the difference between uniaxial PML and stretched-coordinate PML. The calculation reported in the demo are based [on this work
](http://people.ee.duke.edu/~cummer/reprints/138_Popa11_PRA_ComplexTransformationOptics.pdf), where the authors propose a complex coordinate transformation to implement PML for irregularly shaped domain.

In this demo the domain has a regular shape (a square), and therefore what you propose makes sense. However, with `ufl.grad` we are not constrained to any particular shape, and therefore I would say that this approach is "more general". For instance, in [this other PR](https://github.com/FEniCS/dolfinx/blob/a543704a60f7c6869456e34e423b414555ee19a8/python/demo/demo_axis/demo_axis.py#L304-L322) a spherical PML is expressed with the same formulation, and, in principle, it could be used for more complicated shapes as well.